src/engine/fcitx/openbangla.cpp

```diff
@@ -0,0 +1,388 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "openbangla.h"
+#include "keycode.h"
+#include "riti.h"
+#include <fcitx-utils/log.h>
+#include <fcitx-utils/misc.h>
+#include <fcitx-utils/utf8.h>
+#include <fcitx/inputcontext.h>
+#include <fcitx/inputcontextmanager.h>
+#include <fcitx/inputpanel.h>
+#include <memory>
+
+namespace fcitx {
+
+class OpenBanglaState : public InputContextProperty {
+public:
+  OpenBanglaState(OpenBanglaEngine *engine, InputContext &ic)
+      : engine_(engine), ic_(&ic), ctx_(riti_context_new()) {}
+
+  ~OpenBanglaState() {}
+
+  class OpenBanglaCandidate : public CandidateWord {
+  public:
+    OpenBanglaCandidate(OpenBanglaEngine *engine, std::string text, int index)
+        : CandidateWord(Text(std::move(text))), engine_(engine), index_(index) {
+    }
+
+    void select(InputContext *inputContext) const override {
+      auto state = inputContext->propertyFor(engine_->factory());
+      state->selectCandidate(text().toString(), index_);
+    }
+
+  private:
+    OpenBanglaEngine *engine_;
+    int index_;
+  };
+
+  void selectCandidate(const std::string &text, int index) {
+    if (!suggestion_) {
+      return;
+    }
+
+    ic_->commitString(text);
+    riti_context_candidate_committed(ctx_.get(), index);
+    reset();
+  }
+
+  void updatePreedit() {
+    if (!suggestion_) {
+      return;
+    }
+
+    std::string text;
+    if (!riti_suggestion_is_lonely(suggestion_.get())) {
+      auto candidateList = ic_->inputPanel().candidateList();
+      auto index = candidateList->cursorIndex();
+      if (index >= 0 && index < candidateList->size()) {
+        text = candidateList->candidate(index).text().toString();
+      }
+    } else {
+      auto txt = riti_suggestion_get_lonely_suggestion(suggestion_.get());
+      text = txt;
+      // TODO: free txt
+    }
+
+    Text preedit(std::move(text));
+    preedit.setCursor(preedit.textLength());
+    if (ic_->capabilityFlags().test(CapabilityFlag::Preedit)) {
+      ic_->inputPanel().setClientPreedit(preedit);
+      ic_->updatePreedit();
+    }
+    // we don't use input panel preedit here, because the aux text is actually
+    // enough.
+  }
+
+  void commit() {
+    if (!suggestion_) {
+      return;
+    }
+    auto candidateList = ic_->inputPanel().candidateList();
+
+    std::string text;
+    if (!riti_suggestion_is_lonely(suggestion_.get())) {
+      auto candidateList = ic_->inputPanel().candidateList();
+      auto index = candidateList->cursorIndex();
+      if (index >= 0 && index < candidateList->size()) {
+        candidateList->candidate(index).select(ic_);
```

Comment:
I think we have to call `riti_context_candidate_committed(ctx_.get(), index);` as we're committing a candidate.

Reply:
This line will call into OpenBanglaCandidate::select and then enters OpenBanglaState::selectCandidate. So eventually it will call riti_context_candidate_committed